### ISSUE TITLE
Ignore/reconnect for closed connection in pool

### ISSUE THREAD

**Issue by runtel** [aio-libs/aiopg#116]
Hi!

I have some problem with pool.acquire() and closed connection in pool.
Setting on PG Server set connection timeout to 120 second.  When i use pool.acquire() it is raise error:

```
Sleep 600
Traceback (most recent call last):
  File "test.py", line 21, in <module>
    loop.run_until_complete(test_select())
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/base_events.py", line 337, in run_until_complete
    return future.result()
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "test.py", line 10, in test_select
    async with pool.acquire() as conn:
  File "/usr/local/lib/python3.5/site-packages/aiopg/utils.py", line 116, in __aenter__
    self._conn = yield from self._coro
  File "/usr/local/lib/python3.5/site-packages/aiopg/pool.py", line 170, in _acquire
    assert not conn.closed, conn
AssertionError: <aiopg.connection.Connection object at 0x106f7df98>

```

and code
```python
import asyncio
import aiopg

dsn = 'dbname=dbname user=user password=password host=127.0.0.1'
async def test_select():
    async with aiopg.create_pool(dsn) as pool:
        print("Sleep 600")
        await asyncio.sleep(150)
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute("SELECT 1")
                ret = []
                async for row in cur:
                    ret.append(row)
                assert ret == [(1,)]
    print("ALL DONE")


loop = asyncio.get_event_loop()
loop.run_until_complete(test_select())
```

How to solve this problem.
Thanks!

**Comment by runtel** [comment]
Hi!

I override function in class Pool. Connection removed from _free if is closed:
```python
    def _fill_free_pool(self, override_min):
        for i in sorted(range(len(self._free)), reverse=True):
            if self._free[i].closed:
                logger.warning("Connection %s %s is closed and will be removed!!!" % (i, self._free[i]))
                self._free.remove(self._free[i])
        self._cond.notify()
...
```
full code
```python
import asyncio
from aiopg.utils import _PoolContextManager
from aiopg.connection import connect, TIMEOUT
from aiopg.log import logger
from aiopg.pool import Pool

dsn = 'dbname=dbname user=user password=password host=127.0.0.1'

def create_pool(dsn=None, *, minsize=10, maxsize=10,
                loop=None, timeout=TIMEOUT,
                enable_json=True, enable_hstore=True, enable_uuid=True,
                echo=False,
                **kwargs):
    coro = _create_pool(dsn=dsn, minsize=minsize, maxsize=maxsize, loop=loop,
                        timeout=timeout, enable_json=enable_json,
                        enable_hstore=enable_hstore, enable_uuid=enable_uuid,
                        echo=echo, **kwargs)
    return _PoolContextManager(coro)


@asyncio.coroutine
def _create_pool(dsn=None, *, minsize=10, maxsize=10,
                 loop=None, timeout=TIMEOUT,
                 enable_json=True, enable_hstore=True, enable_uuid=True,
                 echo=False,
                 **kwargs):
    if loop is None:
        loop = asyncio.get_event_loop()

    pool = TestPool(dsn, minsize, maxsize, loop, timeout,
                enable_json=enable_json, enable_hstore=enable_hstore,
                enable_uuid=enable_uuid, echo=echo,
                **kwargs)
    if minsize > 0:
        with (yield from pool._cond):
            yield from pool._fill_free_pool(False)
    return pool


class TestPool(Pool):

    @asyncio.coroutine
    def _fill_free_pool1(self, override_min):
        # iterate over free connections and remove timeouted ones
        free_size = len(self._free)
        n = 0
        while n < free_size:
            conn = self._free[-1]
            if conn._reader.at_eof():
                self._free.pop()
                conn.close()
            else:
                self._free.rotate()
            n += 1

        while self.size < self.minsize:
            self._acquiring += 1
            try:
                conn = yield from connect(echo=self._echo, loop=self._loop,
                                          **self._conn_kwargs)
                # raise exception if pool is closing
                self._free.append(conn)
                self._cond.notify()
            finally:
                self._acquiring -= 1
        if self._free:
            return

        if override_min and self.size < self.maxsize:
            self._acquiring += 1
            try:
                conn = yield from connect(echo=self._echo, loop=self._loop,
                                          **self._conn_kwargs)
                # raise exception if pool is closing
                self._free.append(conn)
                self._cond.notify()
            finally:
                self._acquiring -= 1


    @asyncio.coroutine
    def _fill_free_pool(self, override_min):
        for i in sorted(range(len(self._free)), reverse=True):
            if self._free[i].closed:
                logger.warning("Connection %s %s is closed and will be removed!!!" % (i, self._free[i]))
                self._free.remove(self._free[i])
        self._cond.notify()

        while self.size < self.minsize:
            self._acquiring += 1
            try:
                conn = yield from connect(
                    self._dsn, loop=self._loop, timeout=self._timeout,
                    enable_json=self._enable_json,
                    enable_hstore=self._enable_hstore,
                    enable_uuid=self._enable_uuid,
                    echo=self._echo,
                    **self._conn_kwargs)
                # raise exception if pool is closing
                self._free.append(conn)
                self._cond.notify()
            finally:
                self._acquiring -= 1
        if self._free:
            return

        if override_min and self.size < self.maxsize:
            self._acquiring += 1
            try:
                conn = yield from connect(
                    self._dsn, loop=self._loop, timeout=self._timeout,
                    enable_json=self._enable_json,
                    enable_hstore=self._enable_hstore,
                    enable_uuid=self._enable_uuid,
                    echo=self._echo,
                    **self._conn_kwargs)
                # raise exception if pool is closing
                self._free.append(conn)
                self._cond.notify()
            finally:
                self._acquiring -= 1


async def test_select():
    async with create_pool(dsn, echo=True) as pool:
        print("Sleep 150")
        await asyncio.sleep(150)
        conn = await pool.acquire()
        async with conn.cursor() as cur:
            await cur.execute("SELECT 1")
            ret = []
            async for row in cur:
                ret.append(row)
            assert ret == [(1,)]
    print("ALL DONE")


loop = asyncio.get_event_loop()
loop.run_until_complete(test_select())
```

The same solution for aiomysql.
May be exists another solution?

**Comment by jettify** [comment]
Hi @runtel sorry for late reply, just wanted to point to aiomysql pool too :). Would you like to port aiomysql solution and tests in new PR? `self._free.remove()` is `O(n)` that is why `self._free.rotate` is used see more info (https://github.com/aio-libs/aiomysql/pull/27)

**Comment by runtel** [comment]
Hi @jettify 
It works on aiopg and must work on aiomysql
```python
        n, free = 0, len(self._free)
        while n < free:
            conn = self._free[-1]
            if conn.closed:
                self._free.pop()
            else:
                self._free.rotate()
            n += 1
```

I can test and make PR for aiopg.
OK?

**Comment by jettify** [comment]
Sure, just also add test case similar to `aiomysql`s one.

**Comment by runtel** [comment]
All is ready. How i can import aiopg from package, not from system?
How can i run unittest(my first test))))

**Comment by jettify** [comment]
Follow instructions here http://aiopg.readthedocs.io/en/latest/contributing.html.

**Comment by mpaolini** [comment]
Can this be closed now?

**Comment by jettify** [comment]
yes, issue is fixed in #117
